tempodb/blocklist/poller.go

```diff
@@ -456,6 +466,46 @@ func (p *Poller) tenantIndexPollError(idx *backend.TenantIndex, err error) error
 	return nil
 }
 
+// deleteTenant will delete all of a tenant's objects if there is not a tenant index present.
+func (p *Poller) deleteTenant(ctx context.Context, tenantID string) error {
+	level.Info(p.logger).Log("msg", "deleting tenant", "tenant", tenantID)
+
+	if p.cfg.EmptyTenantDeletionAge == 0 {
+		return fmt.Errorf("empty tenant deletion age must be greater than 0")
+	}
+
+	var foundObjects []string
+	err := p.reader.Find(ctx, backend.KeyPath{tenantID}, func(opts backend.FindOpts) {
+		level.Info(p.logger).Log("msg", "checking object for deletion", "object", opts.Key, "modified", opts.Modified)
+
+		if time.Since(opts.Modified) > p.cfg.EmptyTenantDeletionAge {
+			foundObjects = append(foundObjects, opts.Key)
+		}
+	})
+	if err != nil {
+		return err
+	}
+
+	// do nothing if the tenant index has appeared.
+	_, err = p.reader.TenantIndex(ctx, tenantID)
+	// If we have any error other than that which indicates that the tenant index
+	// call was made successfully, and that it does not exist, do nothing.  Only
+	// proceed if we know that the index does not exist.
+	if !errors.Is(err, backend.ErrDoesNotExist) {
+		return nil
+	}
+
+	for _, object := range foundObjects {
+		dir, name := path.Split(object)
```

Comment:
There is some friction here between Find and Delete, where Find returns the path as a single string and Delete requires it split between name and keypath.  Does it make sense for them to match?  Obviously very minor, but something to consider.  Honestly maybe the better approach is to do a cleanup pass over the whole backend interfaces since they have grown closer towards generic filesystem operations after adding List/Find/Delete.

Reply:
Hmm, I see this also.  Looking at `raw.go` , the `Delete` was modeled after `Write` and `Append`.  We could break from this and just receive the string path on `Delete`.

In the case of the find function implementation here, we could collect the objects in a different form than the string path to avoid the split and then rejoin later [here](https://github.com/zalegrala/tempo/blob/aa6308f7627564fcc91c4de5da521297c50e984f/tempodb/backend/s3/s3.go#L248), but I'm not sure.

Reply:
Checking again, all the backends are doing the join on `backend.KeyPath` and `name`.  So how do we feel about the `Delete` call just taking a string path?

Alternatively, we could follow up in another PR to come up with a more holistic approach.

Reply:
That's ok. Since there's not a clear preference for either way, let's leave it.